src/transaction/log_recovery_redo_parallel.cpp

```diff
@@ -0,0 +1,397 @@
+	// spin-wait and try again
+	return nullptr;
+      }
+  }
+
+  void redo_parallel::redo_job_queue::notify_to_be_waited_for_op_finished()
+  {
+    assert (to_be_waited_for_op_in_progress == true);
+    to_be_waited_for_op_in_progress = false;
+    // notify all waiting threads, as there can be many which can pick-up work
+    to_be_waited_for_op_in_progress_cv.notify_all();
+  }
+
+  void redo_parallel::redo_job_queue::notify_in_progress_vpid_finished (VPID _vpid)
+  {
+    bool set_empty = false;
+    {
+      std::lock_guard<std::mutex> lock_in_progress_vpids (in_progress_vpids_mutex);
+      assert (in_progress_vpids.find (_vpid) != in_progress_vpids.cend());
+      in_progress_vpids.erase (_vpid);
+      set_empty = in_progress_vpids.empty ();
+    }
+    if (set_empty)
+      {
+	in_progress_vpids_empty_cv.notify_one();
+      }
+  }
+
+  void redo_parallel::redo_job_queue::wait_for_idle()
+  {
+    {
+      std::unique_lock<std::mutex> empty_queues_lock (produce_queue_mutex);
+      queues_empty_cv.wait (empty_queues_lock, [this]()
+      {
+	return queues_empty;
+      });
+    }
+
+    {
+      std::unique_lock<std::mutex> empty_in_progress_vpids_lock (in_progress_vpids_mutex);
+      in_progress_vpids_empty_cv.wait (empty_in_progress_vpids_lock, [this]()
+      {
+	return in_progress_vpids.empty ();
+      });
+    }
+  }
+
+  /**********************
+   * redo_parallel::redo_task
+   **********************/
+
+  constexpr unsigned short redo_parallel::redo_task::WAIT_AND_CHECK_MILLIS;
+
+  redo_parallel::redo_task::redo_task (std::size_t a_task_id,
+				       redo_parallel::redo_task_active_state_bookkeeping &a_task_active_state_bookkeeping,
+				       redo_job_queue &a_queue)
+    : task_id (a_task_id), task_active_state_bookkeeping (a_task_active_state_bookkeeping), queue (a_queue)
+  {
+    // important to set this at this moment and not when execution begins
+    // to circumvent race conditions where all tasks haven't yet started work
+    // while already bookkeeping is being checked
+    task_active_state_bookkeeping.set_active (task_id);
+
+    log_zip_realloc_if_needed (undo_unzip_support, LOGAREA_SIZE);
+    log_zip_realloc_if_needed (redo_unzip_support, LOGAREA_SIZE);
+  }
+
+  redo_parallel::redo_task::~redo_task()
+  {
+    log_zip_free_data (undo_unzip_support);
+    log_zip_free_data (redo_unzip_support);
+  }
+
+  void redo_parallel::redo_task::execute (context_type &context)
+  {
+    bool finished = false;
+
+    //
+    context.tran_index = LOG_SYSTEM_TRAN_INDEX;
+
+    for (; !finished ;)
+      {
+	bool adding_finished = false;
+	ux_redo_job_base job = queue.locked_pop (adding_finished);
+
+	if (job == nullptr && adding_finished)
+	  {
+	    finished = true;
+	  }
+	else
+	  {
+	    if (job == nullptr)
+	      {
+		// TODO: if needed, check if requested to finish ourselves
+
+		// expecting more data, sleep and check again
+		std::this_thread::sleep_for (std::chrono::milliseconds (WAIT_AND_CHECK_MILLIS));
```

Comment:
I will see whether there's a better solution than this `sleep`.

Reply:
In a separate patch, the sleep loop may be replaced with a conditional wait_for.

Reply:
Added http://jira.cubrid.org/browse/LETS-60